lib/SPIRVProducerPass.cpp

```diff
@@ -6127,3 +6127,40 @@ void SPIRVProducerPass::WriteSPIRVBinary() {
     }
   }
 }
+
+bool SPIRVProducerPass::IsTypeNullable(const Type* type) const {
+  switch (type->getTypeID()) {
+    case Type::HalfTyID:
+    case Type::FloatTyID:
+    case Type::DoubleTyID:
+    case Type::IntegerTyID:
+    case Type::VectorTyID:
+      return true;
+    case Type::PointerTyID: {
+      const PointerType *pointer_type = cast<PointerType>(type);
+      if (pointer_type->getPointerAddressSpace() !=
+          AddressSpace::UniformConstant) {
+        auto pointee_type = pointer_type->getPointerElementType();
+        if (pointee_type->isStructTy() &&
+            cast<StructType>(pointee_type)->isOpaque()) {
+          // Images and samplers are not nullable.
+          return false;
+        }
+      }
+      return true;
+    }
+    case Type::ArrayTyID:
+      return IsTypeNullable(cast<CompositeType>(type)->getTypeAtIndex(0u));
+    case Type::StructTyID: {
+      const StructType* struct_type = cast<StructType>(type);
+      // Images and samplers are not nullable.
+      if (struct_type->isOpaque()) return false;
+      for (const auto element : struct_type->elements()) {
+        if (!IsTypeNullable(element)) return false;
+      }
+      return true;
+    }
+    default:
+      return false;
```

Comment:
I suspect some compilers will not do the control flow analysis to see that all paths lead to a return of a bool.
I would code this as default: break; and then return false as last statement in the function.

But let's wait until we see that problem.